src/DsaBottomUp.cc

```diff
@@ -50,10 +50,104 @@ static const Value *findUniqueReturnValue(const Function &F) {
   return onlyRetVal;
 }
 
+template <typename Set>
+static void markReachableNodes(const Node *n, Set &set) {
+  if (!n)
+    return;
+  assert(!n->isForwarding() && "Cannot mark a forwarded node");
+
+  if (set.insert(n).second)
+    for (auto const &edg : n->links())
+      markReachableNodes(edg.second->getNode(), set);
+}
+
+template <typename Set>
+static void reachableNodes(const Function &fn, Graph &g, Set &inputReach,
+                           Set &retReach) {
+  // formal parameters
+  for (const Value &arg : fn.args()) {
+    if (g.hasCell(arg)) {
+      Cell &c = g.mkCell(arg, Cell());
+      markReachableNodes(c.getNode(), inputReach);
+    }
+  }
+
+  // globals
+  for (auto &kv : g.globals())
+    markReachableNodes(kv.second->getNode(), inputReach);
+
+  // return value
+  if (g.hasRetCell(fn))
+    markReachableNodes(g.getRetCell(fn).getNode(), retReach);
+}
+
+// Copy (recursively) all callsites owned by calleeC to graph callerG.
+// Precondition: callerC simulates calleeC.
+static void copyCallSites(Graph &calleeG, const Cell &calleeC, Graph &callerG,
```

Comment:
I don't understand the logic. Seems way to complex.

First we clone a graph `G_src` into `G_dst` using cloner `C`. Then, given a call site `call` in `G_src`, if `cell(call)` is known to cloner `C`, copy `call` to `G_dst` and set `cell(call) = C(cell(call))` in `G_dst`.

No recursion, everything is already handled by the cloner.

Reply:
To use the cloner was my first idea. When we copy callsites from callee to caller we need to copy cell(call) from the callee into the caller. However, the Cloner only understands nodes. When I copy cell(call) I need a *valid reference* to the cell. I cannot get that from the cloner. That's the motivation of having `copyCallSites`. Essentially, it mimics the traversal that the Cloner does but it has always a valid reference to cells that I can use when I copy a call site. 
`copyCallSites` is also the reason why `Node` has also allocation sites. To make it efficient, when a traverse all reachable cells I just ask each cell's node for the allocation sites that I should copy. I can also get that information by asking the callee's graph but it would be less efficient.

Reply:
You probably mean 'callsites' instead of 'allocation sites' at the end of the comment. Otherwise, I don't understand it at all.

A cell `C` is a pair node `N` and offset `off`. If you know how to map nodes, you know how to map cells. Cloner maps cells all the time because nodes have links and links are cells.

Reply:
Yes, I meant callsite.